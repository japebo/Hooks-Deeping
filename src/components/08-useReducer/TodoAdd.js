import React from 'react'
import { useForm } from '../../hooks/useForm';


export const TodoAdd = ({handleAdd}) => {
    const [ formValue, handleInputChange, reset ] = useForm({
        task: ''
    });

    const handleSubmit = (e) => {
        e.preventDefault();
        
        if ( formValue.task.trim().length <= 1 ) {
            reset();
            return;
        }
        
        handleAdd( {
            id: new Date().getTime(),
            task: formValue.task,
            done: false    
        })

        reset();
    }

    return (
        <>
            <h4> Add Task </h4>

            <form onSubmit={ handleSubmit }>
                <input
                    type="text"
                    name="task"
                    className='form-control'
                    placeholder="Aprender..."
                    value={ formValue.task }
                    autoComplete="off"
                    onChange={ handleInputChange }
                />
                <button
                    type='submit'
                    className='btn btn-outline-primary mt-1 col-12'
                >
                    Add
                </button>
            </form>
        </>
    )
}
