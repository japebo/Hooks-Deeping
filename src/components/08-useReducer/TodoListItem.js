import React from 'react'

export const TodoListItem = ({todo, index, handleDeletion, handleToggle}) => {
    return (
        <li
            key={todo.id}
            className='list-group-item'
        >
            <p 
                className={ `${todo.done===true && 'complete'}` }
                onClick={ () => {handleToggle(todo.id)} } > {index+1}. {todo.task} </p>
            <button 
                className='btn btn-danger'
                onClick={ () => { handleDeletion(todo.id) } }
            >
                Delete
            </button>
        </li>
    )
}
