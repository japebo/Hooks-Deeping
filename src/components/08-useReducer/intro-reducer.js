
const initialState = [{
    id: 1,
    task: 'Botar la basura',
    done: false
}];

const todoReducer = ( state = initialState, action) => {

    if (action?.type === 'add') {
        return [...state, action.payload]
    }

    return state;
}

let todos = todoReducer();

const newTask = {
    id: 2,
    task: 'Buscar el pan',
    done: false
};

const actionToPerform = {
    type: 'add',
    payload: newTask
};

todos = todoReducer( todos, actionToPerform );

console.log(todos);