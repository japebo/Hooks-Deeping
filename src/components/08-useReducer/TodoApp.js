import React, { useEffect, useReducer } from 'react'
import './styles.css'
import { TodoAdd } from './TodoAdd';
import { TodoList } from './TodoList';
import { todoReducer } from './todoReducer';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];

    // return [{
    //     id: new Date().getTime(),
    //     task: 'Aprender React',
    //     done: false
    // }];
}

export const TodoApp = () => {

    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem( 'todos', JSON.stringify(todos));
    }, [todos])


    const handleAdd = (newTodo) => {
        dispatch({
            type: 'add',
            payload: newTodo
        });
    }

    const handleDeletion = ( todoID ) => {
        const action = {
            type: 'delete',
            payload: todoID
        }

        dispatch(action);

    }

    const handleToggle = ( todoID ) => {

        dispatch({
            type: 'toggle',
            payload: todoID
        });

    }

    return (
        <div>
            <h1>ToDoApp ( {todos.length} )</h1>
            <hr/>

            <div className='row'>
                <div className='col-7'>
                    < TodoList todos={todos} handleDeletion={handleDeletion} handleToggle={handleToggle}/>
                </div>

                <div className='col-5'>
                    <TodoAdd handleAdd={handleAdd}/>
                </div>
            </div>
        </div>
    )
}
