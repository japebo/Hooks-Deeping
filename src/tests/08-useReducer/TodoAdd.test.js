import { shallow } from "enzyme"
import { TodoAdd } from "../../components/08-useReducer/TodoAdd"

describe('Preubas en <TodoAdd/>', () => {
    const handleAdd = jest.fn();
    const wrapper = shallow(<TodoAdd handleAdd={handleAdd}/>);

    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();  
    })

    test('No debe llamar la función handleAdd', () => {
        const formSubmit = wrapper.find('form').prop('onSubmit');
        formSubmit({preventDefault(){}}); //forma corta de definir una función (shorthand method definition). Este forma de crear las funciones puede ser utilizado como método en la declaración de un objeto o en las clases de ES6.
        
        expect(handleAdd).toHaveBeenCalledTimes(0);
    })
    
    test('Debe llamar la función handleAdd', () => {
        const value = 'Aprender React';
        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'task'
            }});
            const formSubmit = wrapper.find('form').prop('onSubmit');
            formSubmit({preventDefault(){}}); //forma corta de definir una función (shorthand method definition). Este forma de crear las funciones puede ser utilizado como método en la declaración de un objeto o en las clases de ES6.
            expect(handleAdd).toHaveBeenCalledTimes(1);
            // expect(handleAdd).toHaveBeenCalledWith(expect.any(Object));
            expect(handleAdd).toHaveBeenCalledWith({
                id: expect.any(Number),
                task: value,
                done: false
            });
            console.log(wrapper.find('input').text());
            expect(wrapper.find('input').prop('value')).toBe('');
        })
        
})
