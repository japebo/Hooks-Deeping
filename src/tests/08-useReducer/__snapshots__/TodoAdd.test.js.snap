// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preubas en <TodoAdd/> Debe mostrarse correctamente 1`] = `
<Fragment>
  <h4>
     Add Task 
  </h4>
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="task"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <button
      className="btn btn-outline-primary mt-1 col-12"
      type="submit"
    >
      Add
    </button>
  </form>
</Fragment>
`;
